### PR TITLE
Bug fix/internal issue #738

### PR DESCRIPTION
**Pull request by Dronplane**
This PR adds abstraction layer for processing arguments of ArangoSearch functions (currently for PHRASE). 
Filter could be built now not only from AstNodes but from evaluated VPackSlice also. This makes possible to use dynamically composed parameters (but still deterministic) for PHRASE function.
So queries like this will be possible:
`FOR t IN wikipedia LIMIT 10  LET token = {WILDCARD: CONCAT(TOKENS(t.title, 'tokenizer')[0], "%")}
 FOR d in v_wiki_phrase SEARCH PHRASE(d.title, token, 'tokenizer') 
   RETURN d`

This is done by introducing ArgsTraits class which holds functions to work with provided arguments: getting values, enumerating members.

https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/11269/

### CODE REVIEW

**Comment by Dronplane** [comment]
tests blue except known failure

**Comment by gnusi** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11372/

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 1]
```diff
@@ -2140,13 +2320,12 @@ arangodb::Result fromFuncPhraseLike(char const* funcName,
                                     char const* subFuncName,
                                     irs::by_phrase* filter,
                                     QueryContext const& ctx,
-                                    arangodb::aql::AstNode const& elem,
+                                    VPackSlice const& elem,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2323]:
```suggestion
                                    VPackSlice elem,
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 2]
```diff
@@ -2033,13 +2033,208 @@ arangodb::Result fromFuncMinMatch(
...
+        return SCOPED_VALUE_TYPE_OBJECT;
+      case VPackValueType::Null:
+        return SCOPED_VALUE_TYPE_NULL;
+      default:
+        break; // Make Clang happy
+    }
+    return SCOPED_VALUE_TYPE_INVALID;
+  }
+
+  static VPackSlice const& valueSlice(ValueType const& v) noexcept {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2136]:
```suggestion
  static ValueType valueSlice(ValueType v) noexcept {
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 3]
```diff
@@ -2033,13 +2033,208 @@ arangodb::Result fromFuncMinMatch(
...
+      return true;
+    }
+    return false;
+  }
+
+  constexpr static bool isDeterministic(VPackSlice const&) {
+    return true;
+  }
+
+  static size_t numMembers(VPackSlice const& arg) {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2166]:
```suggestion
  static size_t numMembers(VPackSlice const arg) {
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 4]
```diff
@@ -2033,13 +2033,208 @@ arangodb::Result fromFuncMinMatch(
...
+
+  static bool getValueString(ValueType const& v, irs::string_ref& str) {
+    if (v.isString()) {
+      str = ::getStringRef(v);
+      return true;
+    }
+    return false;
+  }
+
+  constexpr static bool isDeterministic(VPackSlice const&) {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2162]:
```suggestion
  constexpr static bool isDeterministic(VPackSlice) {

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 5]
```diff
@@ -2033,13 +2033,208 @@ arangodb::Result fromFuncMinMatch(
...
+  }
+
+  static size_t numMembers(VPackSlice const& arg) {
+    if (arg.isArray()) {
+      return arg.length();
+    }
+    return 1;
+  }
+
+  static arangodb::Result getMemberValue(VPackSlice const& arg, size_t idx,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2173]:
```suggestion
  static arangodb::Result getMemberValue(VPackSlice arg, size_t idx,
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 6]
```diff
@@ -2033,13 +2033,208 @@ arangodb::Result fromFuncMinMatch(
...
+                                         bool, QueryContext const&, bool&) {
+    TRI_ASSERT(arg.isArray());
+    TRI_ASSERT(arg.length() > idx);
+    value = arg.at(idx);
+    return {};
+  }
+
+  template<typename T>
+  static arangodb::Result evaluateArg(T& out, ValueType& value, char const* funcName,
+    VPackSlice const& args, size_t i, bool isFilter, QueryContext const& ctx) {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2184]:
```suggestion
    VPackSlice args, size_t i, bool isFilter, QueryContext const& ctx) {
```

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 7]
```diff
@@ -2033,13 +2033,208 @@ arangodb::Result fromFuncMinMatch(
...
+};
+
 typedef std::function<
   arangodb::Result(char const*,
                    size_t const,
                    char const*,
                    irs::by_phrase*,
                    QueryContext const&,
-                   arangodb::aql::AstNode const&,
+                   VPackSlice const&,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2237]:
```suggestion
                   VPackSlice,

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 8]
```diff
@@ -2052,35 +2247,22 @@ std::string getSubFuncErrorSuffix(char const* funcName, size_t const funcArgumen
 arangodb::Result oneArgumentfromFuncPhrase(char const* funcName,
                                            size_t const funcArgumentPosition,
                                            char const* subFuncName,
-                                           irs::by_phrase* filter,
-                                           QueryContext const& ctx,
-                                           arangodb::aql::AstNode const& elem,
-                                           ScopedAqlValue& termValue,
+                                           VPackSlice const& elem,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2250]:
```suggestion
                                           VPackSlice elem,

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 9]
```diff
@@ -2090,13 +2272,12 @@ arangodb::Result fromFuncPhraseTerm(char const* funcName,
                                     char const* subFuncName,
                                     irs::by_phrase* filter,
                                     QueryContext const& ctx,
-                                    arangodb::aql::AstNode const& elem,
+                                    VPackSlice const& elem,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2275]:
```suggestion
                                    VPackSlice elem,

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 10]
```diff
@@ -2285,7 +2465,7 @@ arangodb::Result fromFuncPhraseLevenshteinMatch(char const* funcName,
                                                 char const* subFuncName,
                                                 irs::by_phrase* filter,
                                                 QueryContext const& ctx,
-                                                arangodb::aql::AstNode const& array,
+                                                VPackSlice const& array,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2468]:
```suggestion
                                                VPackSlice array,

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 11]
```diff
@@ -2546,7 +2696,7 @@ arangodb::Result fromFuncPhraseInRange(char const* funcName,
                                        char const* subFuncName,
                                        irs::by_phrase* filter,
                                        QueryContext const& ctx,
-                                       arangodb::aql::AstNode const& array,
+                                       VPackSlice const& array,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2699]:
```suggestion
                                       VPackSlice array,

File: arangod/IResearch/IResearchFilterFactory.cpp [hunk 12]
```diff
@@ -2621,139 +2759,138 @@ std::map<std::string, ConversionPhraseHandler> const FCallSystemConversionPhrase
...
   {"IN_RANGE", fromFuncPhraseInRange}
 };
-
-// {<TERM>|<STARTS_WITH>|<WILDCARD>|<LEVENSHTEIN_MATCH>|<TERMS>|<IN_RANGE>: '[' <term> [, ...] ']'}
 arangodb::Result processPhraseArgObjectType(char const* funcName,
                                             size_t const funcArgumentPosition,
                                             irs::by_phrase* filter,
                                             QueryContext const& ctx,
-                                            arangodb::aql::AstNode const& object,
+                                            VPackSlice const& object,
```
Comment (gnusi) [inline arangod/IResearch/IResearchFilterFactory.cpp:2769]:
```suggestion
                                            VPackSlice object,

**Comment by Dronplane** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/11377/

**Pull request merged**